torch/csrc/autograd/engine.cpp

```diff
@@ -630,7 +630,10 @@ void Engine::evaluate_function(
     if (auto* capture_vec = fn_info.captures_.get()) {
       // Lock mutex for writing to graph_task->captured_vars_.
       std::lock_guard<std::mutex> lock(graph_task->mutex_);
-      for (auto capture : *capture_vec) {
+      for (const auto& capture : *capture_vec) {
+        for (auto& hook : capture.hooks_) {
+          inputs[capture.input_idx_] = (*hook)(inputs[capture.input_idx_]);
```

Comment:
What is the justification to modify the input to the next Node here?
Shouldn't this only modify the captured value?

Reply:
It's a contract similar to what a chain of function pre-hooks does. 
Which returned value to capture if there are multiple hooks, if a hook only modifies the captured grad? Only the last one counts?

Reply:
For the captured value I agree: the returned value should be the one captured.
But for the rest of the graph and the following Nodes I'm not so sure.